### PR TITLE
L2 -> L1 synchronization thread

### PR DESCRIPTION
**Pull request by m-kus**

### CODE REVIEW

File: crates/client/settlement/src/lib.rs
Comment (tdelabro) [inline crates/client/settlement/src/lib.rs:45]: Does every field have to be public?

File: crates/client/settlement/src/lib.rs
Comment (tdelabro) [inline crates/client/settlement/src/lib.rs:52]: I believe StarkFelt impl default. So you can probably derive Default

File: crates/client/settlement/src/mockup/mod.rs
Comment (tdelabro) [inline crates/client/settlement/src/mockup/mod.rs:37]: forgotten println

File: crates/client/settlement/src/errors.rs
Comment (tdelabro) [inline crates/client/settlement/src/errors.rs:11]: what does `level` means in this context?

File: crates/client/settlement/src/errors.rs
Comment (tdelabro) [inline crates/client/settlement/src/errors.rs:30]: Where are the recoverable errors defined?

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:39]: DO you really want to `return` here? Isn't it more a place where we want to `continue`?

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:50]: Maybe the signature should be `-> !` and this should be a `panic!`. Wdyt?

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:58]: what is the meaning of `_proc` here?

Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:58]: Can you use a more expressive error type here? An enum.

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:115]: Can we move this in some sort of client/utils crate?

Comment (tdelabro) [reply crates/client/settlement/src/sync_state.rs:115]: There is probably code to share with the RPCs

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:202]: We are short-circuiting the SNOS run here, isn't it?

Can you add a comment about it please somewhere

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:162]: doc

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:133]: doc

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:145]: doc

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:58]: doc

File: starknet-rpc-test/src/lib.rs
Comment (tdelabro) [inline starknet-rpc-test/src/lib.rs:190]: Silence them back before merging

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/settlement/src/sync_state.rs
Comment (bidzyyys) [inline crates/client/settlement/src/sync_state.rs:37]: Is type coercion for `Arc` not working here?

File: .gitignore
Comment (bidzyyys) [inline .gitignore:22]: Why do you need logs here? Is it possible to log somewhere else?

File: starknet-rpc-test/src/lib.rs
Comment (m-kus) [reply starknet-rpc-test/src/lib.rs:190]: I've added a flag to `MadaraClient::init` to enable/disable logging, I suppose this might be useful for other scenarios as well. I will set it to false everywhere, also can make it feature enabled if you'd like

File: crates/client/settlement/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/mod.rs:25]: But they are both foreign?

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs:22]: Yeah, I think I'll move to a dedicated error type instead of anyhow, it will make things simpler

File: crates/client/settlement/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/mod.rs:25]: @tdelabro do you think we could add this cast to starknet-api?

File: crates/client/data-availability/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/data-availability/src/ethereum/mod.rs]: After some thoughts I created a separate client for Starknet state contract

File: starknet-rpc-test/src/lib.rs
Comment (m-kus) [reply starknet-rpc-test/src/lib.rs:188]: I made logs optional, now they can be enabled by providing MADARA_LOG env variable which should contain a relative log folder, is that ok? Generally I think there might be other cases when you'd want to check Madara logs for troubleshooting (at least I couldn't debug efficiently without them)

Comment (m-kus) [reply starknet-rpc-test/src/lib.rs:193]: Added readme for integration tests

File: crates/client/settlement/src/sync_state.rs
Comment (m-kus) [reply crates/client/settlement/src/sync_state.rs:40]: Inside sync_state_proc we have another loop which listens for the block finalization events, it will exit when the finality_notifications_stream will be closed/terminated -- I assume upon the node termination?

File: crates/client/settlement/src/sync_state.rs
Comment (m-kus) [reply crates/client/settlement/src/sync_state.rs:51]: Wow, thanks, I didn't know about this type!

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: The Starkness level where state root differs from what we got from the settlement contract

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: Now it's more explicitly defined

File: crates/client/settlement/src/sync_state.rs
Comment (m-kus) [reply crates/client/settlement/src/sync_state.rs]: It's inner loop, renamed

File: crates/client/settlement/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/mod.rs]: It will fail and outer loop will retry unless the error is not recoverable

File: crates/client/settlement/src/lib.rs
Comment (m-kus) [reply crates/client/settlement/src/lib.rs:34]: Typically I'd agree, but in this very case we get the SNOS output as a result of a previous conversion and we don't need it for any other purpose.

File: crates/client/settlement/src/mockup/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/mockup/mod.rs:37]: We could but there's no use for it, at least in the current scope

File: crates/client/settlement/src/sync_state.rs
Comment (m-kus) [reply crates/client/settlement/src/sync_state.rs]: I tried to reuse as much as possible, also here we rely on crate-level error type (now)

File: crates/client/settlement/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/mod.rs:15]: Let's refactor afterwards, I guess we will have more things to reuse, e.g. Starknet contract bindings

Comment (tdelabro) [reply crates/client/settlement/src/ethereum/mod.rs:15]: Yep. Keep it like this for now. In the future we will have the common felt type with hopefully conversion methods to u256

File: crates/node/src/service.rs
Comment (tdelabro) [inline crates/node/src/service.rs:428]: This souldn't be required as you are doing  match? You get an ettor?

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: It's `<B as Block>::Hash`, but we would need to make Error generic then

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs:24]: Yes, chain height, I'll rename

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: Yes, I'll rename

File: crates/client/settlement/src/mockup/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/mockup/mod.rs]: It's fixed in the latest revision, chain_id is retrieved from the runtime

File: crates/client/settlement/src/ethereum/mod.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/mod.rs]: It's likely an outdated version, anyhow is no longer used

File: crates/node/src/service.rs
Comment (m-kus) [reply crates/node/src/service.rs:428]: Since mockup is not feature-enabled, I marked the enum as #[non_exhaustive] (afaik this is considered as good practice in such cases). Thus we need to handle _

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs:38]: It's a generic type, which means we either have to make our error generic as well, or lock to a specific ethers middleware (provider + wallet type)

File: starknet-rpc-test/settlement_worker.rs
Comment (m-kus) [reply starknet-rpc-test/settlement_worker.rs:1]: Sounds good!

File: crates/client/settlement/src/errors.rs
Comment (tdelabro) [reply crates/client/settlement/src/errors.rs]: Sound ok for me to make it generic. As this error is clearly linked to the substrate framework

File: crates/client/settlement/src/errors.rs
Comment (tdelabro) [reply crates/client/settlement/src/errors.rs:38]: Okay. But use `thiserror::Error` instead. And let the display impl do the conversion to string

File: crates/node/src/service.rs
Comment (tdelabro) [reply crates/node/src/service.rs:428]: https://doc.rust-lang.org/reference/attributes/type_system.html

It seems that it is used to guarantee backward compatibility downstream. I don't think this is something we care about.

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: Not for the particular settlement provider implementations, e.g. for Ethereum contract

Comment (tdelabro) [reply crates/client/settlement/src/errors.rs]: The error message do mention Substrate: `    #[error("Failed to find Substrate block header for hash: {0}")]`
So I assume it will always be Substrate related

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/settlement/src/errors.rs
Comment (m-kus) [reply crates/client/settlement/src/errors.rs]: This seems outdated, pls check the latest commit

File: crates/client/settlement/src/ethereum/client.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/client.rs]: Also outdated

File: crates/client/settlement/Cargo.toml
Comment (m-kus) [reply crates/client/settlement/Cargo.toml:14]: For which crates?

Comment (tdelabro) [reply crates/client/settlement/Cargo.toml:14]: As much as possible. All those from madara primitive. Probably also those from `sp`. And a bunch of other.
It's rare that a crate does not have default features

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:38]: Do you have to move the value? You are only using it by reference

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:94]: I don't like that we are using this in order to sync l2.
Different reason:
1) https://crates.parity.io/sc_client_api/client/trait.BlockchainEvents.html#tymethod.finality_notification_stream the doc state their is no guarantee an event is fired
2) If this thread goes down all the network node has to go done, otherwise we may miss some events

I would rather use a combination of https://crates.parity.io/sp_blockchain/trait.Backend.html#tymethod.last_finalized (in order to get the last finalized one) and https://crates.parity.io/sp_blockchain/trait.HeaderBackend.html#tymethod.header (to get every block from last_settled to last_finalized). This way this thread can be down, we can reboot the node and it will still correctly sync

Comment (tdelabro) [reply crates/client/settlement/src/sync_state.rs:94]:
```rust
fn sync() -> !  {
let last_settled = get_last_settled();
let last_finalized = get_last_finalized();

  if last_settled.number == last_finalized.number {
     sleep.await;
     continue; 
  }
  if last_settled.number > last_finalized.number {
     panic!();
  }

  for block_number in last_settled.number..last_finalized.number {
     let block = get_block(block_number);
     <do our thing>
  }

}
```

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:170]: Is this value going to change once the settlement contract is deployed? If no, we can query it once at the beginning of the thread and use it forever after

Comment (tdelabro) [reply crates/client/settlement/src/sync_state.rs:170]: Maybe it's an assumption we don't want to do?

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [inline crates/client/settlement/src/sync_state.rs:189]: `height` rather than `level`

**Review by tdelabro — CHANGES_REQUESTED** [review]
I think you should review the design of the worker

File: crates/client/settlement/src/ethereum/client.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/client.rs:87]: It can potentially be useful for troubleshooting in the future

File: crates/client/settlement/src/sync_state.rs
Comment (m-kus) [reply crates/client/settlement/src/sync_state.rs:103]: Regarding the second item - this is handled in the current implementation, the thread can catch up with the head.
It also handles the situations where Madara state is behind the settled one (e.g. you are bootstrapping a new node) - in your snipped you panic, which is prob not what you'd want.

Overall, listening the finality notification stream is basically the same as I would poll last_finalized() method. Given that we cannot miss events, it's probably ok even if it's not guaranteed that event is not fired, as long as there will be another one next.

File: crates/client/settlement/src/ethereum/client.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/client.rs]: I'm now using anyhow to bypass original error, as we discussed, is it ok?

File: crates/client/settlement/src/ethereum/client.rs
Comment (m-kus) [reply crates/client/settlement/src/ethereum/client.rs]: Reduced to debug level

File: crates/client/settlement/src/sync_state.rs
Comment (tdelabro) [reply crates/client/settlement/src/sync_state.rs:103]: Okay, I missed that part. So you fallback yo the solution I described if the chain is not up to date.
Seems okay

**Comment by m-kus** [comment]
Superseded by #1279

**Pull request closed without merging**
